File: [0300][Medium][Longest_Increasing_Subsequence]/Longest_Increasing_Subsequence.py
# leetcode time     cost : 1220 ms
# leetcode memory   cost : 13.8 MB 
# Time  Complexity: O(n*k)
# Space Complexity: O(n)
# slolution 3, DP
class Solution:
    #def lengthOfLIS(self, nums: List[int]) -> int:
    def lengthOfLIS(self, nums):    
        if not nums:
            return 0
        
        n = len(nums)
        dp = [1] * n

        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], 1 + dp[j])
                    
        return max(dp)
    
def main():
    nums = [1,3,6,7,9,4,10,5,6] #expect is 6
    obj = Solution()
    result = obj.lengthOfLIS(nums)
    print("return result is "+str(result));

if __name__ =='__main__':
    main() 

File: [0300][Medium][Longest_Increasing_Subsequence]/Longest_Increasing_Subsequence.java
// leetcode time     cost : ---
// leetcode memory   cost : --- 
// Time  Complexity: O(2**n)
// Space Complexity: O(n**2)
// slolution 1, brute force recursion
public class Solution {

    public int lengthOfLIS(int[] nums) {
        return lengthofLIS(nums, Integer.MIN_VALUE, 0);
    }

    public int lengthofLIS(int[] nums, int prev, int curpos) {
        if (curpos == nums.length) {
            return 0;
        }
        int taken = 0;
        if (nums[curpos] > prev) {
            taken = 1 + lengthofLIS(nums, nums[curpos], curpos + 1);
        }
        int nottaken = lengthofLIS(nums, prev, curpos + 1);
        return Math.max(taken, nottaken);
    }
}

public class Longest_Increasing_Subsequence {
    public static void main(String args[]) {
        int[] nums = {1,3,6,7,9,4,10,5,6}  ;   // expect is 6
        Solution Solution_obj = new Solution();
        int result = Solution_obj.lengthOfLIS(nums);
        System.out.println("In java code,return value is :" + (result));
    }
}